Added user deletion docs (#58)
danielpaul/RapidRails: docs/user_deletion.md

> # User Deletion

Reviewer: so we need to highlight that there are three things that can happen or we have built for: 

we need to mention that they should choose the option based on their terms and conditions and privacy policy of the app that they are building. 

1. just soft delete the user (default if no rake task for annon is setup) - no data is deleted
2. soft delete and after a period of time annon the data of the user so there is no PII (personal identifiable information). say how they can set this to be 0 so it happens immediately or how to customise the time duration. 
3. just delete the user permanently immediately when the user requests and how they would do this.

# User Deletion

In it's default setup, deleting a user account revokes their access to the application and soft-deletes the User record. We use the [discard](https://github.com/jhawthorn/discard) gem to soft-delete records across the application.

Depending on legalities of your application, you can choose from three options:

1. Soft-delete(default)
2. Soft-delete and anonymization
3. Permanentaly delete the user record

## How to delete a user account

1. A signed in user goes to their **Account Settings** and clicks the **Delete Account** button, enters their password and optional feedback and then click the **Delete Account Now** button. This will soft-delete the record and sign them out.

## User Anonymization and Customization

> We have an optional rake task that if setup, anonymizes user records after a period of time. This task calls `AnonymizationService` which strips all columns of the user's data.

Reviewer: link to the file in our GitHub repo so they can click on the AnonymizationService and view the file if they would like to

Reviewer: same for other things we mention in the code. like below for the variable that they need to set. we want to make it easy for users to jump to code places without them having to go and have a look at things themselves.

1. Run `rake anonymize:users` to anonymize user records that were discarded a certain number of days ago and have not been anonymized yet. The number of days is stored in a constant called `ANONYMIZE_USER_DATA_AFTER_DAYS` which if set to **0** will anonymize the user record immediately.

> 2. Some fields are set to custom values. For example, `full_name` to **"Deleted User"**, `email` to **"prefix@domain"** where **prefix** and **domain** are customizable and password is set to a secure 60 character **SecureRandom** password.

Reviewer: we need to mention here that they need to update the script that if they add new fields and models that store PII, then they need to update the script to randomise those as well. won't happen automatically. 

should be maybe in a important highlighted heading so it draws attention to it.

# User Deletion

In it's default setup, deleting a user account revokes their access to the application and soft-deletes the User record. We use the [discard](https://github.com/jhawthorn/discard) gem to soft-delete records across the application.

Depending on legalities of your application, you can choose from three options:

1. Soft-delete(default)
2. Soft-delete and anonymization

> 3. Permanentaly delete the user record

Reviewer: so for each of these we need to have a heading under named the same so they can look at that section for what they need to do for each. 

otherwise at the moment it seems a bit too all over the place where we need to read the whole doc to see what the user actually needs to do. 

in each of the sections:
1. say they don't need to do anything and this is how ti works
2. mention about the task to setup and what it does and how they should update the script
3. how they would set this up

Reviewer: basically the same content that we have below but organised it better so its clear

# User Deletion

In it's default setup, deleting a user account revokes their access to the application and soft-deletes the User record. We use the [discard](https://github.com/jhawthorn/discard) gem to soft-delete records across the application.

A signed in user can delete their account via the **Account Settings**. They may enter optional feedback which is stored as a new [UserAccountFeedback](../app/models/user_account_feedback.rb) record.

![](images/delete_account.png)

Depending on legalities of your application, you can choose from three options:

> 1. Soft-delete(default)

Reviewer: ```suggestion
1. Soft-delete (default)
```

> Depending on the way your app would like to handle user deletion, you can setup your app for any of the following or write your own flow:

Reviewer: ```suggestion
Depending on the way your app would like to handle user deletion, you can setup your app for any of the following or write your own flow  - based on your app's terms and conditions:
```

2. Soft-delete and anonymization
3. Permanentaly delete the user record

> ## Soft-delete (default)

Reviewer: ```suggestion
## 1. Soft-delete (default)
```

> The default flow soft-deletes the User record and sign them out.

Reviewer: ```suggestion
The default flow soft-deletes the User record and sign them out. The user will not be able to sign up again with the same email and would need to contact your support to restore their account (or permanently delete/anonymize their account).
```

> ## Soft-delete and anonymization

Reviewer: ```suggestion
## 2. Soft-delete and anonymization
```

> We have a task `rake anonymize:users` that anonymizes user records after a period of time. This task calls [AnonymizationService](../app/services/anonymization_service.rb) which strips all columns of the user's data.

Reviewer: ```suggestion
This is useful where you want to get rid of the user's data from your database but only after a period of time. A use case would be if you still want to retain some data for support queries, refunds, etc. after the user deletes their account and then get rid of the data later. 

We have a task `rake anonymize:users` that anonymizes user records after a period of time. This task calls [AnonymizationService](../app/services/anonymization_service.rb) which strips all columns of the user's data.
```

1. Running `rake anonymize:users` anonymizes user records that were discarded a certain number of days ago and have not been anonymized yet. The number of days is stored in a constant called [ANONYMIZE_USER_DATA_AFTER_DAYS](../config/initializers/0_constants.rb) which if set to **0** will anonymize the user record immediately (when the rake task runs). The rake task should be setup on your cloud platform to run periodically if needed.
2. Some fields are set to custom values. For example, `full_name` to **"Deleted User"**, `email` to **"prefix@domain"** where **prefix** and **domain** are customizable and password is set to a secure 60 character **SecureRandom** password.

### IMPORTANT:

> If using the anonymizer and if any new attributes are added to the **User** record, don't forget to update the method `anonymize_user` in [AnonymizationService](../app/services/anonymization_service.rb) to clear or randomize the new attributes.

Reviewer: ```suggestion
If using the anonymizer for all the new attributes and new models that store personal information, you need to update the method `anonymize_user` in [AnonymizationService](../app/services/anonymization_service.rb) to clear or randomize those values.
```
